tests/jax/test_custom_partitioning.py

```diff
@@ -0,0 +1,301 @@
+        devices = np.asarray(jax.devices()[:DEVICE_COUNT]).reshape(*mesh_shape)
+        with global_shard_guard(_get_sharding_resource(mesh_names, sharding_type)[0]):
+            with jax.sharding.Mesh(devices, mesh_names):
+                x_ = random.normal(subkeys[0], qkv_shape)
+                bias_ = random.normal(subkeys[1], bias_shape)
+                mask_ = make_decoder_mask(tokens)
+                graded_f = jax.value_and_grad(func, argnums=(0, 1, 2))
+                pjitter = pjit(graded_f)
+                out = {}
+                hlo = pjitter.lower(x_, bias_, mask_).compile().as_text()
+                dic = count_collective(hlo)
+                assert dic==collective_ref, f"Expected number of collective is: {dic==collective_ref=}, but got {dic=}."
+
+class TestCPGenerator:
+    @pytest.mark.parametrize('mesh_shape,mesh_names,sharding_type, collective_ref', MESH_CONFIG)
+    @pytest.mark.parametrize('input_shape', [(32, 128)])
+    @pytest.mark.parametrize('other_shape', [(128,)])
+    @pytest.mark.parametrize('batch_dim', [0])
+    @pytest.mark.parametrize('zero_centered_gamma', [False])
+    @pytest.mark.skipif(not is_devices_enough(DEVICE_COUNT), reason='Num of GPU is not enough')
+    def test_layernorm(self, mesh_shape, mesh_names, sharding_type, collective_ref, input_shape, other_shape,
+                       batch_dim, zero_centered_gamma):
+        
+        def func(x, gamma, beta):
+            x = layernorm(x, gamma, beta, 'layernorm', zero_centered_gamma, epsilon)
+            return jnp.mean(x)
+       
+        x_ = random.normal(random.PRNGKey(1124), input_shape)
+        gamma = jnp.ones(other_shape)
+        beta = jnp.ones(other_shape)
+        x_spec, gamma_spec, beta_spec = _get_sharding_spec(mesh_names, sharding_type)
+        devices = np.asarray(jax.devices()[:DEVICE_COUNT]).reshape(*mesh_shape)
+        mesh = jax.sharding.Mesh(devices, mesh_names)
+        with mesh, global_shard_guard_new(_get_sharding_resource(mesh_names, sharding_type)[1]):
+            x_ = jax.device_put(x_, NamedSharding(mesh, x_spec))
+            gamma = jax.device_put(gamma, NamedSharding(mesh, gamma_spec))
+            beta = jax.device_put(beta, NamedSharding(mesh, beta_spec))
+            graded_f = jax.value_and_grad(func, argnums=(0, 1, 2))
+            pjitter = pjit(graded_f,
+                         in_shardings=[x_spec, gamma_spec, beta_spec],
+                        out_shardings=(None, (x_spec, gamma_spec, beta_spec),))
+                
+            hlo = pjitter.lower(x_, gamma, beta).compile().as_text()
+            dic = count_collective(hlo)
+            test_l, test_grads = pjitter(x_, gamma, beta)
+            dic = count_collective(hlo)
+            assert dic==collective_ref, f"Expected number of collective is: {collective_ref=}, but got {dic=}."
+
+    @pytest.mark.parametrize('mesh_shape,mesh_names,sharding_type, collective_ref', MESH_CONFIG)
```

Comment:
This test for softmax with xla-based sharding is showing 1 all-reduce for me.

I would expect softmax to not have any collective ops only if the tensor is sharded so that each device has the entire tensor in the dimension aligned with the exponential sum. That doesn't seem to be happening here. If we want the softmax test to check for 0 collectives, we'll need to shard the tensor differently here, right?

Reply:
We need to test many different input sharding and for each, we need an hardcoded list of expected collective.